layer1/transcation/src/transaction.mligo

```diff
@@ -0,0 +1,90 @@
+(* MIT License
+
+   Copyright (c) 2022 Marigold <contact@marigold.dev>
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to deal in
+   the Software without restriction, including without limitation the rights to
+   use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+   the Software, and to permit persons to whom the Software is furnished to do so,
+   subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in all
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE. *)
+
+#import "../../stdlib_ext/src/result.mligo" "Stdlib_Result"
+#import "../../commons/transaction.mligo" "Tx"
+
+(** transaction lib *)
+
+(** [Result] module provides info on failure *)
+module Result = struct
+  (** transaction errors *)
+  type error
+    = Not_enough_balance
+    | Non_existed_arg
+    | Not_implement_yet
```

Comment:
I think there is a question, if the transited statue is incorrect, do we want to record this error in state?
So users can know why their balance doesn't get changed.

Reply:
Oh I see, it is a kind of "diagnosis" sent by the L2?

Reply:
Well for the balance, it is useful, as it can be a reason for the verifier to reject a proof. But for the argument, it really depend on the VM:
- contract loading opcode
- argument reading opcode
- etc.
`Non_existed_arg` could be the correct result of the `READ_ARGUMENT` opcode for example. However, it is not necessarily an error of the verifier, it can be legitimate result of the execution. It is only an error of the verifier if the proof pretended that it worked.
Diagnosis is good, but if it comes to the one-step verifier, then it is a disagreement between two nodes on the state of the rollup. So the question is "is it a proof of the correct result", not "did my transaction went ok".

Reply:
I think that depends on the definition of our `state`.

correct proof could be
- my transaction went ok
- my transaction went south expectedly.
like
```
type state = Either ledger (Error e)
```

current implementation is more like
```
type state = ledger
```

I vote `type state = Either ledger (Error e)`, what do you think?